src/proposal.mligo

```diff
@@ -341,6 +363,68 @@ let delete_proposal
+let to_signed(n : unsigned_quorum_fraction): quorum_fraction
+  = { numerator = int(n.numerator) }
+
+[@inline]
+let to_unsigned(n : quorum_fraction): unsigned_quorum_fraction
+  = { numerator = match is_nat(n.numerator) with
+              | Some n -> n
+              | None -> (failwith("BAD_UNSIGNED_CONVERSION"):nat)
+    }
+
+let update_quorum(store, config : storage * config): storage =
+  let current_period = get_current_period_num(store.start_time, config.voting_period) in
+  let current_cycle = period_to_cycle(current_period) in
+  if store.quorum_threshold_at_cycle.last_updated_cycle = current_cycle
+    then store // Quorum has been updated in this period, so no change is required.
+    else
+      if current_cycle > store.quorum_threshold_at_cycle.last_updated_cycle
+          then
+              let previous_staked = store.quorum_threshold_at_cycle.staked in
+              let previous_participation = to_signed(fraction_to_quorum_fraction(previous_staked, config.governance_total_supply)) in
+              let old_quorum = to_signed(store.quorum_threshold_at_cycle.quorum_threshold) in
+              let quorum_change = to_signed(config.quorum_change) in
+              let possible_new_quorum =
+                // old_quorum + (previous_participation - old_quorum) * quorum_change
+                fadd(old_quorum, fmul(quorum_change, fsub(previous_participation, old_quorum))) in
+                //old_quorum + fmul(quorum_change, previous_participation - old_quorum) in
+              let one_plus_max_change_percent = to_signed({ numerator = config.max_quorum_change.numerator + quorum_denominator }) in
+              let min_new_quorum =
+                fdiv(old_quorum, one_plus_max_change_percent) in
+              let max_new_quorum =
+                fmul(old_quorum, one_plus_max_change_percent) in
+
+              let config_min_qt = to_signed(config.min_quorum_threshold) in
+              let config_max_qt = to_signed(config.max_quorum_threshold) in
+              let new_quorum = bound_qt(bound_qt(possible_new_quorum, min_new_quorum, max_new_quorum), config_min_qt, config_max_qt)
+              in { store with quorum_threshold_at_cycle =
+                   { quorum_threshold = to_unsigned(new_quorum)
+                   ; last_updated_cycle = current_cycle
+                   ; staked = 0n;
+                   }
+                 }
+            else store
+
+let propose (param, config, store : propose_params * config * storage): return =
+  let store = check_is_proposal_valid (config, param, store) in
+  let store = check_proposal_limit_reached (config, store) in
+  let amount_to_freeze = param.frozen_token + config.fixed_proposal_fee_in_token in
+  let qt = store.quorum_threshold_at_cycle.quorum_threshold in
+    // We store the current quorum threshold before it gets updated in the following update_quorum call,
+    // and use the current quorum to initiate the proposal object.
+  let store = update_quorum(store, config) in
```

Comment:
Why did you add this, shouldn't the quorum be updated to the new cycle here exactly so that the new value is used?

Beside even if this wasn't the case this would work only for the first `propose` called in a cycle, not the following ones.

Reply:
What a stupid mixup! Have fixed it now and added a test.